src/Microsoft.AspNetCore.Authentication/AuthenticationBuilder.cs

```diff
@@ -84,6 +90,17 @@ public AuthenticationBuilder(IServiceCollection services)
             return AddScheme<TOptions, THandler>(authenticationScheme, displayName, configureOptions: configureOptions);
         }
 
+        /// <summary>
+        /// Adds a <see cref="VirtualAuthenticationHandler"/> based authentication handler which can be used to 
+        /// redirect to other authentication schemes.
+        /// </summary>
+        /// <param name="authenticationScheme">The name of this scheme.</param>
+        /// <param name="displayName">The display name of this scheme.</param>
+        /// <param name="configureOptions">Used to configure the scheme options.</param>
+        /// <returns>The builder.</returns>
+        public virtual AuthenticationBuilder AddVirtualScheme(string authenticationScheme, string displayName, Action<VirtualSchemeOptions> configureOptions)
```

Comment:
Extension?

Reply:
The other methods were public virtual so I just did the same for consistency...  This also seems slightly easier to customize if someone wanted to return their own derived custom AuthenticationBuilder from an AddXyz() that changed the behavior of `AddVirtualScheme`

Reply:
I still don't see why virtual is getting special treatment here, from the outside it's not mechanically different from any other auth handler. Let's start with an extension method like the other providers.

Reply:
Its not special, the other handler in the core package (RemoteAuthenticationHandler) is already also a virtual methods on the builder as opposed to an extension, this is just continuing that pattern.

Reply:
We could have written AddScheme itself as an extension method, at this point I think any handlers we add in core Authentication should go on the AuthenticationBuilder as a virtual, while anything added in another package would be an extension method (unless they are returning a custom AuthenticationBuilder in their `AddXyz()`).  I think the VirtualScheme is more of a core concept like RemoteAuth, unlike something narrow and specific like a cookie/oauth handler

Reply:
I would actually almost argue that we should consider moving VirtualScheme into the HttpAbstractions real core auth layer, but we aren't setup to do that easily right now...